<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.269">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Preceptor’s Primer for Bayesian Data Science: Using the Cardinal Virtues for Inference - 3&nbsp; Data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./04-rubin-causal-model.html" rel="next">
<link href="./02-wrangling.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }"><div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Preceptor’s Primer for Bayesian Data Science: Using the Cardinal Virtues for Inference</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/PPBDS/primer/" title="Source Code" class="sidebar-tool px-1"><i class="bi bi-github"></i></a>
    <a href="" title="Download" id="sidebar-tool-dropdown-0" class="sidebar-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi bi-download"></i></a>
    <ul class="dropdown-menu" aria-labelledby="sidebar-tool-dropdown-0">
<li>
          <a class="dropdown-item sidebar-tools-main-item" href="./Preceptor-s-Primer-for-Bayesian-Data-Science--Using-the-Cardinal-Virtues-for-Inference.pdf">
            <i class="bi bi-bi-file-pdf pe-1"></i>
          Download PDF
          </a>
        </li>
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="./Preceptor-s-Primer-for-Bayesian-Data-Science--Using-the-Cardinal-Virtues-for-Inference.epub">
            <i class="bi bi-bi-journal pe-1"></i>
          Download ePub
          </a>
        </li>
    </ul>
</div>
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Welcome</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./preamble.html" class="sidebar-item-text sidebar-link">Preamble</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./00-getting-started.html" class="sidebar-item-text sidebar-link">Getting Started</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-visualization.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Visualization</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-wrangling.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Wrangling</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03-data.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04-rubin-causal-model.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Rubin Causal Model</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-probability.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Probability</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-one-parameter.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">One Parameter</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-two-parameters.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Two Parameters</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08-three-parameters.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Three Parameters</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09-four-parameters.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Four Parameters</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10-five-parameters.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Five Parameters</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11-n-parameters.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">N Parameters</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./set-up.html" class="sidebar-item-text sidebar-link">Set Up for Working on The Primer</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./style-guide.html" class="sidebar-item-text sidebar-link">Style Guide</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
</div>
</nav><!-- margin-sidebar --><div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="toc-section-number">3.1</span>  Introduction</a></li>
  <li>
<a href="#reading-and-writing-files" id="toc-reading-and-writing-files" class="nav-link" data-scroll-target="#reading-and-writing-files"><span class="toc-section-number">3.2</span>  Reading and writing files</a>
  <ul class="collapse">
<li><a href="#text-files" id="toc-text-files" class="nav-link" data-scroll-target="#text-files"><span class="toc-section-number">3.2.1</span>  Text files</a></li>
  <li><a href="#excel-files" id="toc-excel-files" class="nav-link" data-scroll-target="#excel-files"><span class="toc-section-number">3.2.2</span>  Excel files</a></li>
  <li><a href="#rds-files" id="toc-rds-files" class="nav-link" data-scroll-target="#rds-files"><span class="toc-section-number">3.2.3</span>  RDS files</a></li>
  <li><a href="#json" id="toc-json" class="nav-link" data-scroll-target="#json"><span class="toc-section-number">3.2.4</span>  JSON</a></li>
  </ul>
</li>
  <li>
<a href="#databases" id="toc-databases" class="nav-link" data-scroll-target="#databases"><span class="toc-section-number">3.3</span>  Databases</a>
  <ul class="collapse">
<li><a href="#reading-data-from-a-sqlite-database" id="toc-reading-data-from-a-sqlite-database" class="nav-link" data-scroll-target="#reading-data-from-a-sqlite-database"><span class="toc-section-number">3.3.1</span>  Reading data from a SQLite database</a></li>
  <li><a href="#interacting-with-sqlite-databases" id="toc-interacting-with-sqlite-databases" class="nav-link" data-scroll-target="#interacting-with-sqlite-databases"><span class="toc-section-number">3.3.2</span>  Interacting with SQLite databases</a></li>
  </ul>
</li>
  <li><a href="#webscraping" id="toc-webscraping" class="nav-link" data-scroll-target="#webscraping"><span class="toc-section-number">3.4</span>  Webscraping</a></li>
  <li><a href="#working-with-apis" id="toc-working-with-apis" class="nav-link" data-scroll-target="#working-with-apis"><span class="toc-section-number">3.5</span>  Working with APIs</a></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="toc-section-number">3.6</span>  Summary</a></li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/PPBDS/primer/edit/main/03-data.qmd" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-data" class="quarto-section-identifier d-none d-lg-block"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><p><em>You can never look at the data too much.</em> – Mark Engerman</p>
<section id="introduction" class="level2" data-number="3.1"><h2 data-number="3.1" class="anchored" data-anchor-id="introduction">
<span class="header-section-number">3.1</span> Introduction</h2>
<!-- Any inspiration to be drawn from this? https://ubc-dsci.github.io/introduction-to-datascience/reading.html -->
<!-- Add these?  -->
<!-- https://www.rstudio.com/resources/webinars/part-1-easy-ways-to-collect-different-types-of-data-from-the-web-with-r/ -->
<!-- https://www.rstudio.com/resources/webinars/part-2-easy-ways-to-collect-different-types-of-data-from-the-web-with-r/ -->
<!-- Add Google sheets example. -->
<p>Getting data into and out of R is a major part of any real world data science project.</p>
<p>Load the packages which we will need in this chapter.</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dbplyr.tidyverse.org/">dbplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sfirke/janitor">janitor</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can find most of the files that we use for this chapter <a href="https://github.com/PPBDS/all.primer.tutorials/raw/master/inst/tutorials/033-data-files/data/">here</a>. We can also access these files by saving the URL to R and then using it so that we can download and access the files located on the internet.</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">github_url</span> <span class="op">&lt;-</span> <span class="st">"https://raw.githubusercontent.com/PPBDS/all.primer.tutorials/master/inst/tutorials/033-data-files/data/"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="reading-and-writing-files" class="level2" data-number="3.2"><h2 data-number="3.2" class="anchored" data-anchor-id="reading-and-writing-files">
<span class="header-section-number">3.2</span> Reading and writing files</h2>
<p>This section reviews common file formats. We will review how to pull data from these files into R and how cto reate new files so that we can share data with other people.</p>
<section id="text-files" class="level3" data-number="3.2.1"><h3 data-number="3.2.1" class="anchored" data-anchor-id="text-files">
<span class="header-section-number">3.2.1</span> Text files</h3>
<p>The most common type of data text file is “CSV,” which stands for <strong>c</strong>omma <strong>s</strong>eparated <strong>v</strong>alue. In other words, CSV files are files whose values are separated by commas. Each comma from the csv file corresponds to a column, and the column names are, by default, taken from the first line of the file.</p>
<p>CSV files (and their counterparts) are easily transferable between computers and programming languages and are extremely simple because they’re effectively just text files that have a special format.</p>
<p>However, due to their simple nature, CSV files are only able to move basic data and only as text values. They also have poor support for special characters like commas, which can make the dataset harder to organize and understand by adding a new column for a few specific entries. This make CSV files good for sharing data between computers and languages, but not efficient for transporting or saving large amounts of data.</p>
<section id="reading-and-writing-from-a-csv-file" class="level4" data-number="3.2.1.1"><h4 data-number="3.2.1.1" class="anchored" data-anchor-id="reading-and-writing-from-a-csv-file">
<span class="header-section-number">3.2.1.1</span> Reading and writing from a CSV file</h4>
<p>Here’s an example of a CSV file and what it looks like. Use <code><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv()</a></code> from the <strong>readr</strong> package — which is one of the main packages within the <strong>tidyverse</strong> collection of packages — to load the data into R. The <code>file</code> argument is the file path for the CSV file.</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># We can access files either by using their URL or their file path. In this</span></span>
<span><span class="co"># case, we're using the GitHub URL to access the database. This is done by</span></span>
<span><span class="co"># pasting the file name into our URL using the paste0() command.</span></span>
<span></span>
<span><span class="va">file_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">github_url</span>, <span class="st">"test_1.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">file_1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 2 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (3): a, b, c

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
      a     b     c
  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     1     2     3
2     4     5     6</code></pre>
</div>
</div>
<!-- the path argument is deprecated, use file instead -->
<p>Use <code><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv()</a></code> to save a tibble to a csv file. <code><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv()</a></code> has two main arguments: <code>x</code> and <code>file</code>. The <code>x</code> argument is the data set that you want to save. The <code>file</code> argument is the file path to which you want to save the file. The end of the <code>file</code> argument is the name that you want to use for the file.</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">csv_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html">tribble</a></span><span class="op">(</span></span>
<span>  <span class="op">~</span> <span class="va">`a`</span>, <span class="op">~</span> <span class="va">`b`</span>, <span class="op">~</span> <span class="va">`c`</span>,</span>
<span>      <span class="fl">1</span>,     <span class="fl">2</span>,     <span class="fl">3</span>,</span>
<span>      <span class="fl">4</span>,     <span class="fl">5</span>,     <span class="fl">6</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">csv_data</span>, file <span class="op">=</span> <span class="st">"my_csv.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>When we read the csv file again, the data shows up. This is useful for saving information both to share and for your own projects.</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="st">"my_csv.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 2 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (3): a, b, c

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
      a     b     c
  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     1     2     3
2     4     5     6</code></pre>
</div>
</div>
<p>When you want to remove a file from your system, use <code><a href="https://rdrr.io/r/base/files.html">file.remove()</a></code>.</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.remove</a></span><span class="op">(</span><span class="st">"my_csv.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Sometimes, CSV files will be inconveniently formatted. Maybe they have the wrong column names, have information at the top of the file, or have comments interspersed within the file.</p>
</section><section id="skip" class="level4" data-number="3.2.1.2"><h4 data-number="3.2.1.2" class="anchored" data-anchor-id="skip">
<span class="header-section-number">3.2.1.2</span> <code>skip</code>
</h4>
<p>Consider the following csv file: <code>test_2.csv</code>. Here’s what it looks like as a text file:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>Top two rows consist of junk which
we don't care about. Data starts on row 3.
a,b,c
9,8,7
4,5,6</code></pre>
</div>
</div>
<p>As you can see, there is text at the top of this file. Often times information about how data was collected, or other relevant information, is included at the top of the data file. However, <code><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv()</a></code> can’t differentiate between this text and the data that we want to read, causing it to fail and output gibberish.</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># You can also get a csv file by using the URL of the file. This won't work for</span></span>
<span><span class="co"># all file types though.</span></span>
<span></span>
<span><span class="va">file_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">github_url</span>, <span class="st">"test_2.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">file_2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: One or more parsing issues, call `problems()` on your data frame for details,
e.g.:
  dat &lt;- vroom(...)
  problems(dat)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 4 Columns: 1
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (1): Top two rows consist of junk which

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 1
  `Top two rows consist of junk which`      
  &lt;chr&gt;                                     
1 we don't care about. Data starts on row 3.
2 a,b,c                                     
3 9,8,7                                     
4 4,5,6                                     </code></pre>
</div>
</div>
<p>We can use the <code>skip</code> argument to skip the first 2 text lines and allow <code><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv()</a></code> to work.</p>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">file_2</span>,</span>
<span>         skip <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 2 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (3): a, b, c

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
      a     b     c
  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     9     8     7
2     4     5     6</code></pre>
</div>
</div>
<p>Now that we’ve gotten rid of the warnings, let’s look at the other message R sends: the column specification message.</p>
</section><section id="col_types" class="level4" data-number="3.2.1.3"><h4 data-number="3.2.1.3" class="anchored" data-anchor-id="col_types">
<span class="header-section-number">3.2.1.3</span> <code>col_types</code>
</h4>
<p>The column specification message is a message that R sends to tell you what data types it is using for the column.</p>
<p>Data types are the types discussed in Chapter 2 Wrangling, such as characters, factors, integers, and dates. When we use a tibble, each column has to have a specific type of data. In this example, all of the columns have numbers in them. If there are characters in the column, the columns are going to have a character data type.</p>
<p>To get rid of the column specification message, use the <code>col_types()</code> argument and specify the data types. You can do this by just copying the column specification message and putting it as the <code>col_types()</code> argument.</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">file_2</span>,</span>
<span>         skip <span class="op">=</span> <span class="fl">2</span>,</span>
<span>         col_types <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/cols.html">cols</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_double</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                          b <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_double</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                          c <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_double</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
      a     b     c
  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     9     8     7
2     4     5     6</code></pre>
</div>
</div>
<p>You can also change the column arguments so that you get the data type that you want. Take <code>test_7.csv</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">test_7</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">github_url</span>, <span class="st">"test_7.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">test_7</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 2 Columns: 2
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (1): student
dbl (1): grade

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  grade student
  &lt;dbl&gt; &lt;chr&gt;  
1     1 Sam    
2     5 Becca  </code></pre>
</div>
</div>
<p>Let’s try parsing this so that the <code>student</code> column is a factor and the <code>grade</code> column is an integer.</p>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">test_7</span>,</span>
<span>         col_types <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/cols.html">cols</a></span><span class="op">(</span>grade <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_integer</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                          student <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/parse_factor.html">col_factor</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  grade student
  &lt;int&gt; &lt;fct&gt;  
1     1 Sam    
2     5 Becca  </code></pre>
</div>
</div>
<p>By being clever with the columns, we can make our lives easier down the line when we graph the data.</p>
<p>We can also manipulate other arguments for CSV files.</p>
</section><section id="col_names-and-clean_names" class="level4" data-number="3.2.1.4"><h4 data-number="3.2.1.4" class="anchored" data-anchor-id="col_names-and-clean_names">
<span class="header-section-number">3.2.1.4</span> <code>col_names</code> and <code>clean_names()</code>
</h4>
<p>Let’s try changing the column names in the <code>test_3.csv</code> file.</p>
<p>As you can see below, the file doesn’t have any column names, resulting in the first row being considered as the names for the rest of the file.</p>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">file_3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">github_url</span>, <span class="st">"test_3.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">file_3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 1 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (3): 11, 21, 33

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 3
   `11`  `21`  `33`
  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     4     5     6</code></pre>
</div>
</div>
<p>We can fix this by changing the <code>col_names</code> argument.</p>
<div class="cell">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">file_3</span>, col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"b"</span>, <span class="st">"c"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 2 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (3): a, b, c

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
      a     b     c
  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1    11    21    33
2     4     5     6</code></pre>
</div>
</div>
<p>You can also create names automatically by setting <code>col_names</code> to <code>FALSE</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">file_3</span>, col_names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 2 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (3): X1, X2, X3

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
     X1    X2    X3
  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1    11    21    33
2     4     5     6</code></pre>
</div>
</div>
<p>Changing the names of the columns allows for you to call on the columns later when the data is actually a tibble. Setting the column names to something that you can understand makes it much easier to understand your code later on.</p>
<p>But what if we have good column names, they just aren’t formatted correctly? Let’s look at <code>test_4.csv</code> for an example.</p>
<div class="cell">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">file_4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">github_url</span>, <span class="st">"test_4.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">file_4_tibble</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">file_4</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>New names:
Rows: 3 Columns: 4
── Column specification
──────────────────────────────────────────────────────── Delimiter: "," dbl
(4): one powers, Two_Powers...2, 3_Powers, Two_Powers...4
ℹ Use `spec()` to retrieve the full column specification for this data. ℹ
Specify the column types or set `show_col_types = FALSE` to quiet this message.
• `Two_Powers` -&gt; `Two_Powers...2`
• `Two_Powers` -&gt; `Two_Powers...4`</code></pre>
</div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">file_4_tibble</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  `one powers` Two_Powers...2 `3_Powers` Two_Powers...4
         &lt;dbl&gt;          &lt;dbl&gt;      &lt;dbl&gt;          &lt;dbl&gt;
1            1              2          3              2
2            1              4         81              5
3            1              8         27              8</code></pre>
</div>
</div>
<p>The function runs but the column names are not easy to work with. It is possible to access the column by using a backticks (`) around the name.</p>
<div class="cell">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">file_4_tibble</span><span class="op">$</span><span class="va">`one powers`</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1 1 1</code></pre>
</div>
</div>
<p>But that can still cause problems down the line and it’s just annoying to use backticks every time you want a column name. The <code><a href="https://sfirke.github.io/janitor/reference/clean_names.html">clean_names()</a></code> function from the <strong>janitor</strong> package formats the column names so that they follow the <a href="https://en.wikipedia.org/wiki/Snake_case">underscore separated naming convention</a> and are all unique.</p>
<div class="cell">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">file_4_tibble</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://sfirke.github.io/janitor/reference/clean_names.html">clean_names</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  one_powers two_powers_2 x3_powers two_powers_4
       &lt;dbl&gt;        &lt;dbl&gt;     &lt;dbl&gt;        &lt;dbl&gt;
1          1            2         3            2
2          1            4        81            5
3          1            8        27            8</code></pre>
</div>
</div>
<p>These issues are more common when you’re pulling data off of the internet as that’s normally dirty data and can have a lot of columns formatted in weird ways.</p>
<!-- The tutorial includes a lot of arguments like na and comment, but that makes the book feel like it's really long once it's actually put into text form with paragraphs and stuff. -->
</section><section id="na" class="level4" data-number="3.2.1.5"><h4 data-number="3.2.1.5" class="anchored" data-anchor-id="na">
<span class="header-section-number">3.2.1.5</span> <code>na</code>
</h4>
<p>Another feature of <code><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv()</a></code> is the <code>na</code> argument.</p>
<div class="cell">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">file_5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">github_url</span>, <span class="st">"test_5.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">file_5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 2 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (1): b
dbl (2): a, c

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
      a b         c
  &lt;dbl&gt; &lt;chr&gt; &lt;dbl&gt;
1     1 .         3
2     4 5         6</code></pre>
</div>
</div>
<p><code>test_5.csv</code> is missing a value, and it uses a <code>.</code> as a substitute. This makes the computer think that the period is the actual value of the data point, which is obviously not true (we want numbers instead). By default, <code><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv()</a></code> treats white space like spaces or tabs as a missing value, but you can set this argument directly as well.</p>
<div class="cell">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">file_5</span>,</span>
<span>         na <span class="op">=</span> <span class="st">"."</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 2 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (3): a, b, c

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
      a     b     c
  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     1    NA     3
2     4     5     6</code></pre>
</div>
</div>
</section><section id="comment" class="level4" data-number="3.2.1.6"><h4 data-number="3.2.1.6" class="anchored" data-anchor-id="comment">
<span class="header-section-number">3.2.1.6</span> <code>comment</code>
</h4>
<p>You can also tell the code to ignore comment lines, which may be common in a csv file that was written by a human and as such has comments in it. <code>test_6.csv</code> is a perfect example of this. Here’s what it looks like:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "a,b,c"                          "# This is a comment line"      
[3] "98,99,100"                      "# Here is another comment line"
[5] "4,5,6"                         </code></pre>
</div>
</div>
<p>By setting the <code>comment</code> argument, we’re able to skip lines that have a certain starting point. In this case, the comment is the <code>#</code> sign, so we just need to include that in.</p>
<div class="cell">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">file_6</span>, comment <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 2 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (3): a, b, c

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
      a     b     c
  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1    98    99   100
2     4     5     6</code></pre>
</div>
</div>
<p>If you ever want to skip certain lines, just use the <code>comment</code> argument in order to designate it as something to be skipped. This is best used with cases where the <code><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv()</a></code> command will not compile without it.</p>
</section><section id="read_delim" class="level4" data-number="3.2.1.7"><h4 data-number="3.2.1.7" class="anchored" data-anchor-id="read_delim">
<span class="header-section-number">3.2.1.7</span> <code>read_delim()</code>
</h4>
<p>So far, we’ve covered how we can organize CSV data, but at it’s core we’re working with comma separated values. But what happens when we want to read data from something that doesn’t use commas to separate the values?</p>
<p>When our tabular data comes in a different format, we can use the <code><a href="https://readr.tidyverse.org/reference/read_delim.html">read_delim()</a></code> function instead. For example, a different version of <code>test_6.csv</code> could exist that has no column names and uses pipes (<code>|</code>) as the delimiter instead of commas.</p>
<p>Here’s another file named <code>delim_1</code>, which uses the <code>|</code> to separate the lines instead of a comma like a normal CSV file.</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "population|town"   "150|Cambridge, MA" "92|Newton, MA"    </code></pre>
</div>
</div>
<p>With <code><a href="https://readr.tidyverse.org/reference/read_delim.html">read_delim()</a></code>, we specify the first argument as the path to the file, as done with <code><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv()</a></code>. Then we provide values to the <code>delim</code> argument to have the code use <code>|</code> as the separator instead of the comma.</p>
<div class="cell">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Because delim_1 uses pipes to separate values, we can just use that as our</span></span>
<span><span class="co"># delim value. However, for more complex symbols like tab, we use something</span></span>
<span><span class="co"># different like "\\t". This varies for every symbol, but you can find most</span></span>
<span><span class="co"># delim values on the internet.</span></span>
<span></span>
<span><span class="va">delim_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">github_url</span>, <span class="st">"delim_1.txt"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_delim</a></span><span class="op">(</span><span class="va">delim_1</span>, delim <span class="op">=</span> <span class="st">"|"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 2 Columns: 2
── Column specification ────────────────────────────────────────────────────────
Delimiter: "|"
chr (1): town
dbl (1): population

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  population town         
       &lt;dbl&gt; &lt;chr&gt;        
1        150 Cambridge, MA
2         92 Newton, MA   </code></pre>
</div>
</div>
<p>You can often find CSV files on websites like <a href="https://kaggle.com">kaggle</a> as well as by exporting from an Excel spreadsheet. Keep in mind that data imported off of the internet is often very messy, so try to use some of the functions listed here to clean it up. For a full list of arguments and in-depth documentation about the <code><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv()</a></code> function, please visit this <a href="https://readr.tidyverse.org/reference/read_delim.html">website</a>.</p>
<!-- DK: Put clean_names() example here or wherever you like. DONE-->
</section></section><section id="excel-files" class="level3" data-number="3.2.2"><h3 data-number="3.2.2" class="anchored" data-anchor-id="excel-files">
<span class="header-section-number">3.2.2</span> Excel files</h3>
<p>Excel is a spreadsheet program that use tables to analyze, store, or manipulate data. The tables are composed of cells which include text, numbers, or formulas. Excel files have the filename extensions .xls or .xlsx, and they’re capable of storing additional things that you cannot store in a .csv file such as fonts, text formatting, graphics, etc.</p>
<p>In order to write excel files you have to install complex packages, and they are hard to create. Writing excel files is beyond the scope of this Primer.</p>
<p>This makes Excel files valuable because they’re commonly accepted and usable (Microsoft Excel is a very common program), but they’re also hard to use because you can’t write new data into them. As such, Excel files are common for data that was originally in Excel, like accounting data or other spreadsheet applications.</p>
<p>Reading Excel files is easy. To do so, we use the <code><a href="https://readxl.tidyverse.org/reference/read_excel.html">read_excel()</a></code> function from the <strong>readxl</strong> package.</p>
<div class="cell">
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org">readxl</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Unfortunately, it is not possible to read Excel files directly from the web.</span></span>
<span><span class="co"># So we download the file by hand and then read it in from the current working</span></span>
<span><span class="co"># directory. Note that the "proper" way of handling this would be to create a</span></span>
<span><span class="co"># temp directory with tempdir(), download the file into that directory, read it,</span></span>
<span><span class="co"># and then delete the temp directory. That way, you would not have random</span></span>
<span><span class="co"># downloaded files hanging around.</span></span>
<span></span>
<span><span class="co"># The mode = "wb" is a necessary addition for Windows users because Windows is</span></span>
<span><span class="co"># weird. It's not necessary on MacOS and may cause an error as well.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">github_url</span>, <span class="st">"excel_1.xlsx"</span><span class="op">)</span>, </span>
<span>              destfile <span class="op">=</span> <span class="st">"example_excel.xlsx"</span>, mode <span class="op">=</span> <span class="st">"wb"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html">read_excel</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">"example_excel.xlsx"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
      a     b     c
  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     1     2     3
2     4     5     6</code></pre>
</div>
</div>
<p>If the .xlsx file has multiple sheets, you have to use the <code>sheet</code> argument to specify the sheet number or name. The <code><a href="https://readxl.tidyverse.org/reference/read_excel.html">read_excel()</a></code> function also has arguments similar to the <code><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv()</a></code> function such as <code>col_names</code>, <code>col_types</code>, and <code>na</code>.</p>
<!-- ### Google Sheets -->
<!-- AG: This is a preliminary Primer section based on the googlesheets4 documentation. Shyamsai can change this as she sees fit, but I just wanted something there while the tutorials are created. Plus I started work on this way back when and I needed to push my PUMS changes, so I just commented it out. It's basically just a reword of the documentation though so please change it. -->
<!-- [Google Sheets](https://www.google.com/sheets/about/) is a cloud spreadsheet service that allows you to create and host data on the web in a spreadsheet without needing to use Excel. It's commonly used to store data from [Google Forms](https://www.google.com/forms/about/) and surveys, making it extremely useful for people who are inexperienced with spreadsheet and storage systems. -->
<!-- AG: I think it would be cool to have a communal spreadsheet where people would write in new data about themselves like their name and age. May cause some people to just delete everything though, so I'd like some feedback on this. -->
<!-- When it comes to R, we can both read Google Sheets data and write data into it by using the **googlesheets4** library. -->
<!-- ```{r} -->
<!-- library(googlesheets4) -->
<!-- ``` -->
<!-- This library works by using your Google account to access the sheet, then doing its operations like that. Just run the function `gs4_auth()` in order to load your Google Account information like your email into R. **You do need to cache the account information, as it won't run otherwise.** -->
<!-- If you don't want to interact with the sheet and just want to read from it, you can use `gs4_deauth()` to do things noninteractively. This is useful for websites as they're only compiled once: when you compile it. -->
<!-- ```{r, include=FALSE} -->
<!-- gs4_deauth() -->
<!-- ``` -->
<!-- #### Reading Google Sheets -->
<!-- We can access a sheet in two ways: by using the URL or by using the ID of the sheet. Using the URL is pretty simple: you just use `read_sheet()` -->
<!-- ```{r} -->
<!-- # We're reading an example sheet from the gapminder package located here: -->
<!-- # https://docs.google.com/spreadsheets/d/1U6Cf_qEOhiR9AZqTqS3mbMF3zt2db48ZP5v3rkrAEJY/edit#gid=780868077 -->
<!-- read_sheet("https://docs.google.com/spreadsheets/d/1U6Cf_qEOhiR9AZqTqS3mbMF3zt2db48ZP5v3rkrAEJY/edit#gid=780868077") -->
<!-- ``` -->
<!-- The ID of the sheet is the part of the URL between the "/d" and "/edit", and is often used in the Google Sheets API. You can get the ID and some other basic information about the sheet by running the `gs4_get()` function. -->
<!-- ```{r} -->
<!-- gs4_get("https://docs.google.com/spreadsheets/d/1U6Cf_qEOhiR9AZqTqS3mbMF3zt2db48ZP5v3rkrAEJY/edit#gid=780868077") -->
<!-- id <- "1U6Cf_qEOhiR9AZqTqS3mbMF3zt2db48ZP5v3rkrAEJY" -->
<!-- read_sheet(id) -->
<!-- ``` -->
<!-- And much like Excel files, you can specify a `sheet` argument to set the sheet that you want to pull from. -->
<!-- ```{r} -->
<!-- read_sheet(id, sheet = 2) -->
<!-- ``` -->
<!-- If you just want to access a certain range within a cell, you need to use the `range` argument. -->
<!-- ```{r} -->
<!-- read_sheet(id, range = "Americas!A38:F49") -->
<!-- ``` -->
<!-- Named ranges work as well, but they need to [already be set within your Google Sheet](https://support.google.com/docs/answer/63175?co=GENIE.Platform%3DDesktop&hl=en). -->
<!-- ```{r} -->
<!-- read_sheet(id, range = "canada") -->
<!-- ``` -->
<!-- Additionally, you can skip and limit reading the sheet by utilizing the `skip` and `n_max()` arguments. -->
<!-- ```{r} -->
<!-- read_sheet(id, sheet = "Americas", skip = 37, n_max = 11) -->
<!-- ``` -->
<!-- And if you want to read specific formatting for every cell, just run `sheet_read_cells()`. This creates a row for every single cell and its formatting, allowing you to conduct operations like only reading the highlighted cells. -->
<!-- ```{r} -->
<!-- (df <- range_read_cells(id, range = "canada")) -->
<!-- df$cell[[2]] -->
<!-- ``` -->
<!-- Overall, there's a lot of utility with reading Google Sheets into R, but the utility of Google Sheets truly shines when you write data into them. -->
<!-- #### Writing Google Sheets -->
<!-- You can create new Google Sheets as well with the `gs4_create()` function. This allows you to create new Google Sheets, then share them with another person so that you can work together on the same data set. -->
<!-- ```{r} -->
<!-- ss1 <- gs4_create("sheets-create-demo-1") -->
<!-- ss1 -->
<!-- ``` -->
<!-- You can also add R data frames into Google Sheets. Just list multiple data frames and use it in the `sheets` argument. Of course, you could also just use a vector with the sheet names if you want to have multiple sheets. -->
<!-- ```{r} -->
<!-- ss2 <- gs4_create("sheets-create-demo-2",  -->
<!--                   sheets = list(iris = iris, chickwts = chickwts)) -->
<!-- ss2 -->
<!-- ``` -->
<!-- You can also delete created sheets by using the `googledrive::drive_trash()` function in conjunction with the `gs4_find()` function. -->
<!-- ```{r} -->
<!-- gs4_find("sheets-create-demo") |> -->
<!--   googledrive::drive_trash() -->
<!-- ``` -->
<!-- AG: Going to continue with writing section later. It's based on the documentation listed here: https://googlesheets4.tidyverse.org/articles/articles/write-sheets.html -->
<!-- But what if we actually wanted to create a Google Sheet? -->
<!-- ```{r} -->
<!-- ``` -->
</section><section id="rds-files" class="level3" data-number="3.2.3"><h3 data-number="3.2.3" class="anchored" data-anchor-id="rds-files">
<span class="header-section-number">3.2.3</span> RDS files</h3>
<p>One very important aspect of R is saving objects into RDS files, which store a single R object into a file. These files allow us to save R objects such as plots and tibbles into R and then reload the object that it contains later without re-running the code that made it. This is especially useful for when you’re dealing with bulk data and want to save the plot that comes with it for later so that you don’t have to do all of the data wrangling and plotting again. With a RDS file, you save the entire object, allowing you to do more things with it later without having to go through all of the code. However, RDS files are limited to R projects only, and they’re incomprehensible both to human eyes and to other programming languages. This makes RDS files ideal for saving objects temporarily in your own project or for sharing objects to other R users.</p>
<p>Take the following R object, a graph of the <code>iris</code> data set.</p>
<div class="cell">
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">iris_p</span> <span class="op">&lt;-</span> <span class="va">iris</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sepal.Length</span>, y <span class="op">=</span> <span class="va">Sepal.Width</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Sepal Dimensions of Various Species of Iris"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"Sepal Length"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Sepal Width"</span><span class="op">)</span></span>
<span><span class="va">iris_p</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="03-data_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>When we save to a RDS file, we use the function <code><a href="https://readr.tidyverse.org/reference/read_rds.html">write_rds()</a></code>. Just like <code><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv()</a></code>, this function has two main arguments: <code>x</code> and <code>file</code>. The <code>x</code> argument is the object that you want to save. The <code>file</code> argument is the file path where you want to save the file. This determines the name that you use for the file.</p>
<div class="cell">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_rds.html">write_rds</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">iris_p</span>, file <span class="op">=</span> <span class="st">"iris_p.rds"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code><a href="https://readr.tidyverse.org/reference/read_rds.html">read_rds()</a></code> reads the file back into R. Just like <code><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv()</a></code> <code><a href="https://readr.tidyverse.org/reference/read_rds.html">read_rds()</a></code> has one main argument, which is the path to the file that you are wanting to read into R.</p>
<div class="cell">
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_rds.html">read_rds</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"iris_p.rds"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="03-data_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can then use that R object in more operations, such as adding a trend line.</p>
<div class="cell">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">rds_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_rds.html">read_rds</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"iris_p.rds"</span><span class="op">)</span></span>
<span><span class="va">rds_p</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"loess"</span>,</span>
<span>              formula <span class="op">=</span> <span class="va">y</span> <span class="op">~</span> <span class="va">x</span>,</span>
<span>              se <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="03-data_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>By saving the <code>iris_p</code> plot in a RDS file, we eliminate the time needed to calculate and generate that plot again because we can use the saved information. We can then use the object by reading the file back into r and using it like any normal plot, adding new layers and doing new operations.</p>
<p>There are also .Rdata files which can store multiple objects, but RDS files can accomplish a similar task. This makes it much easier to use RDS files for anything that you want to keep in R.</p>
</section><section id="json" class="level3" data-number="3.2.4"><h3 data-number="3.2.4" class="anchored" data-anchor-id="json">
<span class="header-section-number">3.2.4</span> JSON</h3>
<p>An increasingly common format for sharing data is <em>J</em>avaScript <em>O</em>bject <em>N</em>otation or JSON. Because this format is very general, it is nothing like a spreadsheet. Note that JSON files are often made available via the internet. Several organizations provide a JSON API or a web service that you can connect directly to and from which you can obtain data.</p>
<p>JSON files are a minimal readable format that structures data, and they’re commonly used to transmit data between a server and a web application like a website. For people familiar with the Javascript coding language, you’ll likely see the similarities between the JSON file format and the Javascript syntax. This makes JSON files ideal for internet transport, but they don’t see much use within a project like RDS files do.</p>
<p>The functions <code><a href="https://rdrr.io/pkg/jsonlite/man/fromJSON.html">fromJSON()</a></code> and <code><a href="https://rdrr.io/pkg/jsonlite/man/fromJSON.html">toJSON()</a></code> allow you to convert between R objects and JSON. Both functions come from the <strong>jsonlite</strong> package.</p>
<div class="cell">
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://arxiv.org/abs/1403.2805">jsonlite</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function <code><a href="https://rdrr.io/pkg/jsonlite/man/fromJSON.html">toJSON()</a></code> converts a tibble to JSON format. Consider the <code>example_1</code> tibble:</p>
<div class="cell">
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">example_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>name<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Miguel"</span>, <span class="st">"Sofia"</span>, <span class="st">"Aya"</span>, <span class="st">"Cheng"</span><span class="op">)</span>, </span>
<span>                    student_id <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, exam_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">85</span>, <span class="fl">94</span>, <span class="fl">87</span>, <span class="fl">90</span><span class="op">)</span>, </span>
<span>                    exam_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">86</span>, <span class="fl">93</span>, <span class="fl">88</span>, <span class="fl">91</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">example_1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 4
  name   student_id exam_1 exam_2
  &lt;chr&gt;       &lt;int&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 Miguel          1     85     86
2 Sofia           2     94     93
3 Aya             3     87     88
4 Cheng           4     90     91</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># The pretty argument adds indentation and whitespace when TRUE.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/jsonlite/man/fromJSON.html">toJSON</a></span><span class="op">(</span><span class="va">example_1</span>, pretty <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[
  {
    "name": "Miguel",
    "student_id": 1,
    "exam_1": 85,
    "exam_2": 86
  },
  {
    "name": "Sofia",
    "student_id": 2,
    "exam_1": 94,
    "exam_2": 93
  },
  {
    "name": "Aya",
    "student_id": 3,
    "exam_1": 87,
    "exam_2": 88
  },
  {
    "name": "Cheng",
    "student_id": 4,
    "exam_1": 90,
    "exam_2": 91
  }
] </code></pre>
</div>
</div>
<p>The function <code><a href="https://rdrr.io/pkg/jsonlite/man/fromJSON.html">fromJSON()</a></code> converts JSON format to a tibble.</p>
<div class="cell">
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">json_format_ex</span> <span class="op">&lt;-</span></span>
<span><span class="st">'[</span></span>
<span><span class="st">  {"Name" : "Mario", "Age" : 32, "Occupation" : "Plumber"}, </span></span>
<span><span class="st">  {"Name" : "Peach", "Age" : 21, "Occupation" : "Princess"},</span></span>
<span><span class="st">  {},</span></span>
<span><span class="st">  {"Name" : "Bowser", "Occupation" : "Koopa"}</span></span>
<span><span class="st">]'</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/jsonlite/man/fromJSON.html">fromJSON</a></span><span class="op">(</span><span class="va">json_format_ex</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    Name Age Occupation
1  Mario  32    Plumber
2  Peach  21   Princess
3   &lt;NA&gt;  NA       &lt;NA&gt;
4 Bowser  NA      Koopa</code></pre>
</div>
</div>
<p>Make sure to follow the JSON format exactly when you’re writing JSON files, as the format is what makes them special and what allows them to work.</p>
</section></section><section id="databases" class="level2" data-number="3.3"><h2 data-number="3.3" class="anchored" data-anchor-id="databases">
<span class="header-section-number">3.3</span> Databases</h2>
<div class="cell">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="03-data/images/exploits_of_a_mom.png" class="img-fluid figure-img" width="333"></p>
<p></p><figcaption class="figure-caption">DROP TABLE is a particularly infamous SQL command</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>Databases are one of the most common methods of storing data, as they are capable of storing large amounts of data while also allowing for multiple people to access and change them at the same time. Think of these like a giant book with a bunch of tables that hold data. These are useful by themselves, but with the advent of computers we are now able to use relational databases.</p>
<p>A relational database is a database that has the tables interact with one another based on common data, allowing you to create custom tables from an existing set of records. For example, a relational database may hold multiple tables that use the same ID to keep track of different information, like one table having an ID and a movie name while another has the ID and the rating. You can then combine those two in your code, creating a table with the ID, name, and rating. This allows the person who made the database to easily put in new data as well as allow you to pull out some data at a time without loading the entire database.</p>
<p>It’s common to use and interact with these databases in the real world due to most businesses using relational databases to keep track of their data and to update it at their leisure. It’s not uncommon for databases to hold thousands or even millions of rows due to the need to keep track of data.</p>
<p>In this section, we’ll be going over how to pull data from these databases and how to interact with the data without using the entire database.</p>
<section id="reading-data-from-a-sqlite-database" class="level3" data-number="3.3.1"><h3 data-number="3.3.1" class="anchored" data-anchor-id="reading-data-from-a-sqlite-database">
<span class="header-section-number">3.3.1</span> Reading data from a SQLite database</h3>
<p>SQLite is probably the simplest relational database that one can use in combination with R. SQLite databases are self-contained and usually stored and accessed locally on one computer, instead of on the internet or through the cloud. Data is usually stored in a file with a <code>.db</code> extension. Similar to Excel files, these are not plain text files and cannot be read in a plain text editor.</p>
<p>The first thing you need to do to read data into R from a database is to connect to the database. We do that using <code><a href="https://dbi.r-dbi.org/reference/dbConnect.html">dbConnect()</a></code> function from the <strong>DBI</strong> (database interface) package. This does not read in the data, but simply tells R where the database is and opens up a communication channel.</p>
<div class="cell">
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dbi.r-dbi.org">DBI</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rsqlite.r-dbi.org">RSQLite</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># This example uses a different github URL, so you can't use the paste0() trick</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://github.com/PPBDS/primer/blob/master/03-data/data/can_lang.db?raw=true"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"example_db.db"</span>, mode <span class="op">=</span> <span class="st">"wb"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">con_lang_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dbi.r-dbi.org/reference/dbConnect.html">dbConnect</a></span><span class="op">(</span><span class="fu">RSQLite</span><span class="fu">::</span><span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html">SQLite</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"example_db.db"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Often relational databases have many tables, and their power comes from the useful ways they can be joined. Thus anytime you want to access data from a relational database, you need to know the table names. You can get the names of all the tables in the database using <code><a href="https://dbi.r-dbi.org/reference/dbListTables.html">dbListTables()</a></code>.</p>
<div class="cell">
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tables</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dbi.r-dbi.org/reference/dbListTables.html">dbListTables</a></span><span class="op">(</span><span class="va">con_lang_data</span><span class="op">)</span></span>
<span><span class="va">tables</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "lang"</code></pre>
</div>
</div>
<p>We only get one table name returned, which tells us that there is only one table in this database. To reference a table in the database to do things like select columns and filter rows, we use the <code><a href="https://dplyr.tidyverse.org/reference/tbl.html">tbl()</a></code> function from the <strong>dbplyr</strong> package. The package <strong>dbplyr</strong> allows us to work with data stored in databases as if they were local data frames, which is useful because we can do a lot with big datasets without actually having to bring these vast amounts of data into your computer!</p>
<div class="cell">
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">lang_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/tbl.html">tbl</a></span><span class="op">(</span><span class="va">con_lang_data</span>, <span class="st">"lang"</span><span class="op">)</span></span>
<span><span class="va">lang_db</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># Source:   table&lt;lang&gt; [?? x 6]
# Database: sqlite 3.38.5 [/Users/dkane/Desktop/projects/primer/example_db.db]
   category                              langu…¹ mothe…² most_…³ most_…⁴ lang_…⁵
   &lt;chr&gt;                                 &lt;chr&gt;     &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
 1 Aboriginal languages                  Aborig…     590     235      30     665
 2 Non-Official &amp; Non-Aboriginal langua… Afrika…   10260    4785      85   23415
 3 Non-Official &amp; Non-Aboriginal langua… Afro-A…    1150     445      10    2775
 4 Non-Official &amp; Non-Aboriginal langua… Akan (…   13460    5985      25   22150
 5 Non-Official &amp; Non-Aboriginal langua… Albani…   26895   13135     345   31930
 6 Aboriginal languages                  Algonq…      45      10       0     120
 7 Aboriginal languages                  Algonq…    1260     370      40    2480
 8 Non-Official &amp; Non-Aboriginal langua… Americ…    2685    3020    1145   21930
 9 Non-Official &amp; Non-Aboriginal langua… Amharic   22465   12785     200   33670
10 Non-Official &amp; Non-Aboriginal langua… Arabic   419890  223535    5585  629055
# … with more rows, and abbreviated variable names ¹​language, ²​mother_tongue,
#   ³​most_at_home, ⁴​most_at_work, ⁵​lang_known</code></pre>
</div>
</div>
<p>Although it looks like we just got a data frame from the database, we didn’t! It’s a <em>reference</em>, showing us data that is still in the SQLite database (note the first two lines of the output). It does this because databases are often more efficient at selecting, filtering and joining large data sets than R. And typically, the database will not even be stored on your computer, but rather a more powerful machine somewhere on the web. So R is lazy and waits to bring this data into memory until you explicitly tell it. To do so, we use the <code><a href="https://dplyr.tidyverse.org/reference/compute.html">collect()</a></code> function.</p>
<p>Here we will filter for only rows in the Aboriginal languages category according to the 2016 Canada Census, and then use <code><a href="https://dplyr.tidyverse.org/reference/compute.html">collect()</a></code> to finally bring this data into R as a data frame.</p>
<div class="cell">
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">aboriginal_lang_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">lang_db</span>, <span class="va">category</span> <span class="op">==</span> <span class="st">"Aboriginal languages"</span><span class="op">)</span></span>
<span><span class="va">aboriginal_lang_db</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># Source:   SQL [?? x 6]
# Database: sqlite 3.38.5 [/Users/dkane/Desktop/projects/primer/example_db.db]
   category             language                 mothe…¹ most_…² most_…³ lang_…⁴
   &lt;chr&gt;                &lt;chr&gt;                      &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
 1 Aboriginal languages Aboriginal languages, n…     590     235      30     665
 2 Aboriginal languages Algonquian languages, n…      45      10       0     120
 3 Aboriginal languages Algonquin                   1260     370      40    2480
 4 Aboriginal languages Athabaskan languages, n…      50      10       0      85
 5 Aboriginal languages Atikamekw                   6150    5465    1100    6645
 6 Aboriginal languages Babine (Wetsuwet'en)         110      20      10     210
 7 Aboriginal languages Beaver                       190      50       0     340
 8 Aboriginal languages Blackfoot                   2815    1110      85    5645
 9 Aboriginal languages Carrier                     1025     250      15    2100
10 Aboriginal languages Cayuga                        45      10      10     125
# … with more rows, and abbreviated variable names ¹​mother_tongue,
#   ²​most_at_home, ³​most_at_work, ⁴​lang_known</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">aboriginal_lang_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/compute.html">collect</a></span><span class="op">(</span><span class="va">aboriginal_lang_db</span><span class="op">)</span></span>
<span><span class="va">aboriginal_lang_data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 67 × 6
   category             language                 mothe…¹ most_…² most_…³ lang_…⁴
   &lt;chr&gt;                &lt;chr&gt;                      &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
 1 Aboriginal languages Aboriginal languages, n…     590     235      30     665
 2 Aboriginal languages Algonquian languages, n…      45      10       0     120
 3 Aboriginal languages Algonquin                   1260     370      40    2480
 4 Aboriginal languages Athabaskan languages, n…      50      10       0      85
 5 Aboriginal languages Atikamekw                   6150    5465    1100    6645
 6 Aboriginal languages Babine (Wetsuwet'en)         110      20      10     210
 7 Aboriginal languages Beaver                       190      50       0     340
 8 Aboriginal languages Blackfoot                   2815    1110      85    5645
 9 Aboriginal languages Carrier                     1025     250      15    2100
10 Aboriginal languages Cayuga                        45      10      10     125
# … with 57 more rows, and abbreviated variable names ¹​mother_tongue,
#   ²​most_at_home, ³​most_at_work, ⁴​lang_known</code></pre>
</div>
</div>
<p>Why bother to use the <code><a href="https://dplyr.tidyverse.org/reference/compute.html">collect()</a></code> function? The data looks pretty similar in both outputs shown above. And <strong>dbplyr</strong> provides lots of functions similar to <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> that you can use to directly feed the database reference (i.e.&nbsp;what <code><a href="https://dplyr.tidyverse.org/reference/tbl.html">tbl()</a></code> gives you) into downstream analysis functions (e.g., <code>ggplot2</code> for data visualization and <code>lm</code> for linear regression modeling). However, this does not work in every case; look what happens when we try to use <code>nrow</code> to count rows in a data frame:</p>
<div class="cell">
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">aboriginal_lang_db</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] NA</code></pre>
</div>
</div>
<p>or <code>tail</code> to preview the last 6 rows of a data frame:</p>
<pre><code>tail(aboriginal_lang_db)</code></pre>
<pre><code>## Error: tail() is not supported by sql sources</code></pre>
<p>These functions only work when we use the version that we used collect() on:</p>
<div class="cell">
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">aboriginal_lang_data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 67</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">aboriginal_lang_data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 6
  category             language                  mothe…¹ most_…² most_…³ lang_…⁴
  &lt;chr&gt;                &lt;chr&gt;                       &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
1 Aboriginal languages Tahltan                        95       5       0     265
2 Aboriginal languages Thompson (Ntlakapamux)        335      20       0     450
3 Aboriginal languages Tlingit                        95       0      10     260
4 Aboriginal languages Tsimshian                     200      30      10     410
5 Aboriginal languages Wakashan languages, n.i.…      10       0       0      25
6 Aboriginal languages Woods Cree                   1840     800      75    2665
# … with abbreviated variable names ¹​mother_tongue, ²​most_at_home,
#   ³​most_at_work, ⁴​lang_known</code></pre>
</div>
</div>
<p>In order to delete and stop using an SQLite server, you need to first disconnect the file from the connection be using <code><a href="https://dbi.r-dbi.org/reference/dbDisconnect.html">dbDisconnect()</a></code> and passing in the connection object as an argument. You can then safely delete the database file from your computer by using <code><a href="https://rdrr.io/r/base/files.html">file.remove()</a></code>.</p>
<div class="cell">
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://dbi.r-dbi.org/reference/dbDisconnect.html">dbDisconnect</a></span><span class="op">(</span><span class="va">con_lang_data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.remove</a></span><span class="op">(</span><span class="st">"example_db.db"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Some operations will not work to extract columns or single values from the reference given by the <code>tbl</code> function. Thus, once you have finished your data wrangling of the <code><a href="https://dplyr.tidyverse.org/reference/tbl.html">tbl()</a></code> database reference object, it is advisable to bring it into your local machine’s memory using <code><a href="https://dplyr.tidyverse.org/reference/compute.html">collect()</a></code> as a data frame.</p>
<blockquote class="blockquote">
<p>Warning: Usually, databases are very big! Reading the object into your local machine may give an error or take a lot of time to run so be careful if you plan to do this!</p>
</blockquote>
</section><section id="interacting-with-sqlite-databases" class="level3" data-number="3.3.2"><h3 data-number="3.3.2" class="anchored" data-anchor-id="interacting-with-sqlite-databases">
<span class="header-section-number">3.3.2</span> Interacting with SQLite databases</h3>
<p>Now that we’ve figured out how to get data from a database, let’s look at how to wrangle data within that database.</p>
<p>Because databases normally contain large amounts of data, it’s advisable to do your wrangling before you use <code><a href="https://dplyr.tidyverse.org/reference/compute.html">collect()</a></code> to transform the database table into a tibble. This stops you from pulling large amounts of data onto your computer and just ignoring it.</p>
<p>So let’s try pulling in a database and see how we can manipulate it.</p>
<div class="cell">
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dbi.r-dbi.org">DBI</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rsqlite.r-dbi.org">RSQLite</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Again, use mode = "wb" if you're using a Windows operating system.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://github.com/PPBDS/primer/blob/master/03-data/data/chinook.db?raw=true"</span>,</span>
<span>             dest <span class="op">=</span> <span class="st">"chinook.db"</span>, mode <span class="op">=</span> <span class="st">"wb"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">con</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dbi.r-dbi.org/reference/dbConnect.html">dbConnect</a></span><span class="op">(</span><span class="fu">RSQLite</span><span class="fu">::</span><span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html">SQLite</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"chinook.db"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>First off, look at what data this database holds.</p>
<div class="cell">
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://dbi.r-dbi.org/reference/dbListTables.html">dbListTables</a></span><span class="op">(</span><span class="va">con</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "albums"          "artists"         "customers"       "employees"      
 [5] "genres"          "invoice_items"   "invoices"        "media_types"    
 [9] "playlist_track"  "playlists"       "sqlite_sequence" "sqlite_stat1"   
[13] "tracks"         </code></pre>
</div>
</div>
<p>There are 13 tables in this database. Let’s access the first couple of tables, just so that we can get a good look at them and see how they’re structured.</p>
<div class="cell">
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">albums</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/tbl.html">tbl</a></span><span class="op">(</span><span class="va">con</span>, <span class="st">"albums"</span><span class="op">)</span></span>
<span><span class="va">albums</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># Source:   table&lt;albums&gt; [?? x 3]
# Database: sqlite 3.38.5 [/Users/dkane/Desktop/projects/primer/chinook.db]
   AlbumId Title                                 ArtistId
     &lt;int&gt; &lt;chr&gt;                                    &lt;int&gt;
 1       1 For Those About To Rock We Salute You        1
 2       2 Balls to the Wall                            2
 3       3 Restless and Wild                            2
 4       4 Let There Be Rock                            1
 5       5 Big Ones                                     3
 6       6 Jagged Little Pill                           4
 7       7 Facelift                                     5
 8       8 Warner 25 Anos                               6
 9       9 Plays Metallica By Four Cellos               7
10      10 Audioslave                                   8
# … with more rows</code></pre>
</div>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">artists</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/tbl.html">tbl</a></span><span class="op">(</span><span class="va">con</span>, <span class="st">"artists"</span><span class="op">)</span></span>
<span><span class="va">artists</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># Source:   table&lt;artists&gt; [?? x 2]
# Database: sqlite 3.38.5 [/Users/dkane/Desktop/projects/primer/chinook.db]
   ArtistId Name                
      &lt;int&gt; &lt;chr&gt;               
 1        1 AC/DC               
 2        2 Accept              
 3        3 Aerosmith           
 4        4 Alanis Morissette   
 5        5 Alice In Chains     
 6        6 Antônio Carlos Jobim
 7        7 Apocalyptica        
 8        8 Audioslave          
 9        9 BackBeat            
10       10 Billy Cobham        
# … with more rows</code></pre>
</div>
</div>
<p>As you can see, these tables both have a common column: the <code>ArtistId</code> column. However, they all have different information linked to that ID, such as “albums” having the albums that artist produced and “artists” having the name of that artist.</p>
<p>Here is the <a href="http://schemaspy.org/sample/diagrams/summary/relationships.real.large.png">full relationship diagram</a> of the database that we’re using.</p>
<!-- Let's go over some of the operations that you can do on a SQLite database. -->
<!-- #### Using dbplyr -->

<!-- We've already seen how you can use the **dbpylr** package to get a table from a database, but let's look at how you can also use it to do operations on databases as well. -->
<!-- The **dbpylr** package allows you to use many of the same functions from the **tidyverse** like `select()`, `filter()`, `mutate()` without any issues. -->
<!-- Here's an example of using the **dbpylr** package to get the number of albums each artist has created. -->
<div class="cell">

</div>

<!-- These are the functions that we're already familiar with using, and they work just as well on database tables as well. -->
<!-- However, there are cases where you will have to use SQL code to accomplish certain tasks. -->
<div class="cell">

</div>
<!-- #### Using SQL code -->
<!-- First off, let's try to understand what SQLite and SQL really is. -->
<!-- SQL, or *S*tructured *Q*uery *L*anguage, is a coding language that works to build the relational databases that we've been using so far. SQLite is a special type of database that you can download onto your computer that uses SQL as its basic structure and basic language. Essentially, SQLite uses SQL code to do the things that it does. -->
<!-- While you shouldn't need to write SQL code in this primer, it's very common to read from an SQL database in order to access data so that you can graph it. -->
<!-- SQL runs through something known as a *query*, a line of code that pulls in information from the database. This is effectively asking the database for a specific set of data. This is what's known as a `SELECT` command, which "selects" data from the database and brings it into your computer for your use. -->
<!-- ```{sql, connection = con} -->
<!-- -- This code is written in SQL syntax, so it won't work if you just put it into R.  -->
<!-- -- We'll be going over how to write SQL in R later in this section. -->
<!-- SELECT AlbumId, -->
<!--        Title, -->
<!--        ArtistId -->
<!-- FROM albums; -->
<!-- ``` -->
<!-- You can then modify these queries to do operations like joins. -->
<!-- ```{sql, connection = con} -->
<!-- -- This does the same thing as the inner join from Chapter 2 Wrangling -->

<!-- -- Here we're joining the albums and the artists table by the ArtistId column -->
<!-- SELECT artistid, -->
<!--        NAME, -->
<!--        title -->
<!-- FROM -->
<!--   (SELECT title, -->
<!--           albums.artistid AS "ArtistId", -->
<!--           "name" -->
<!--    FROM albums -->
<!--    INNER JOIN artists ON albums.artistid = artists.artistid); -->
<!-- ``` -->
<!-- >If you're interested in what SQL statements mean and how to write SQL code, visit this [tutorial](https://www.sqlitetutorial.net/). If you're interested in how to create a -->
<!-- database through SQL, visit [here](https://www.datacamp.com/community/tutorials/sqlite-in-r). -->
<!-- In fact, the `dbplyr` commands actually use SQL code to create their effect. -->
<!-- You can run `show_query()` to see the SQL code, or the *query* that `dbplyr` produces. -->
<!-- ```{r} -->
<!-- show_query(band_albums) -->
<!-- ``` -->
<!-- However, you can't do everything from SQL through the `dbplyr` package. For example, you can't use the `slice()` command. -->
<!-- ``` -->
<!-- slice(band_albums, 1:10) -->
<!-- ``` -->
<!-- ``` -->
<!-- ## Error: slice() is not supported on database backends -->
<!-- ``` -->
<!-- This is possible in an SQL database. There are a couple of ways to run SQL code in R, but the easiest way is by using the `dbGetQuery()` function from the **DBI** library. -->
<!-- ```{r} -->
<!-- # This code is just copy-pasted and reformatted from the earlier dbplyr code so -->
<!-- # you can best understand it. -->
<!-- dbGetQuery(con, " -->
<!-- SELECT `artist_id`, -->
<!--        `artist_name`, -->
<!--        `num_albums` -->
<!-- FROM -->
<!--   (SELECT `artistid`, -->
<!--           `name`, -->
<!--           `num_albums`, -->
<!--           `artist_name`, -->
<!--           `artistid` AS `artist_id` -->
<!--    FROM -->
<!--      (SELECT `artistid`, -->
<!--              `name`, -->
<!--              `num_albums`, -->
<!--              `name` AS `artist_name` -->
<!--       FROM -->
<!--         (SELECT `artistid`, -->
<!--                 `name`, -->
<!--                 Count(*) AS `num_albums` -->
<!--          FROM -->
<!--            (SELECT `albumid`, -->
<!--                    `artistid`, -->
<!--                    `title`, -->
<!--                    `name` -->
<!--             FROM -->
<!--               (SELECT `albumid`, -->
<!--                       `title`, -->
<!--                       `LHS`.`artistid` AS `ArtistId`, -->
<!--                       `name` -->
<!--                FROM `albums` AS `LHS` -->
<!--                INNER JOIN `artists` AS `RHS` ON (`LHS`.`artistid` = `RHS`.`artistid`))) GROUP  BY `artistid`))) -->
<!-- WHERE (`num_albums` > 3.0) -->
<!--   ORDER  BY `num_albums` DESC -->
<!-- ") -->
<!-- ``` -->
<!-- This method allows you to only use R code within your project, making it easier. -->
<!-- You can also use SQL code directly in your project by changing the header in an Rmarkdown file. When you code in an Rmarkdown file, you create chunks that look like this: -->
<!-- ``` -->
<!-- ## ```{r} -->
<!-- ## -->
<!-- ## ``` -->
<!-- ``` -->
<!-- You can code in SQL by changing the chunk to look like this: -->
<!-- ``` -->
<!-- ## ```{sql, connection = con, output.var = "your_variable_name"} -->
<!-- ##  -->
<!-- ## ``` -->
<!-- ``` -->
<!-- This allows you to write SQL code in genuine SQL syntax and save the output to a variable. -->
<!-- ``` -->
<!-- ## ```{sql, connection = con, output.var = "your_variable_name"} -->
<!-- ## SELECT * FROM albums -->
<!-- ## ``` -->
<!-- ``` -->
<!-- ```{sql, connection = con, output.var = "your_variable_name", echo = FALSE} -->
<!-- -- SQL uses a semicolon to show the end of the statements -->
<!-- -- However, you don't need to include that for the dbGetQuery() statement -->
<!-- SELECT *  -->
<!-- FROM albums; -->
<!-- ``` -->
<!-- You can then use that variable in your R code. -->
<!-- ```{r} -->
<!-- head(your_variable_name) -->
<!-- ``` -->
<!-- Coding in SQL allows you to do the things that the **dbplyr** package doesn't, such as only using certain rows. If you add `LIMIT 10;` at the end of the query (the first `SELECT` statement), you can limit the query to only the first 10 rows. -->
<!-- ```{sql, connection = con, output.var = "band_albums_sql"} -->
<!-- -- We're writing this code using the sql code chunk method.  -->
<!-- -- This is because it allows you to see what's highlighted and what's not. -->
<!-- -- SQL code is also inconvenient because you can't just add a new line, -->
<!-- -- Instead, you need to treat all of these lines of code as a single statement -->
<!-- -- Then add in the command where you think that it would fit. -->
<!-- SELECT `artist_id`, -->
<!--        `artist_name`, -->
<!--        `num_albums` -->
<!-- FROM -->
<!--   (SELECT `artistid`, -->
<!--           `name`, -->
<!--           `num_albums`, -->
<!--           `artist_name`, -->
<!--           `artistid` AS `artist_id` -->
<!--    FROM -->
<!--      (SELECT `artistid`, -->
<!--              `name`, -->
<!--              `num_albums`, -->
<!--              `name` AS `artist_name` -->
<!--       FROM -->
<!--         (SELECT `artistid`, -->
<!--                 `name`, -->
<!--                 Count(*) AS `num_albums` -->
<!--          FROM -->
<!--            (SELECT `albumid`, -->
<!--                    `artistid`, -->
<!--                    `title`, -->
<!--                    `name` -->
<!--             FROM -->
<!--               (SELECT `albumid`, -->
<!--                       `title`, -->
<!--                       `LHS`.`artistid` AS `ArtistId`, -->
<!--                       `name` -->
<!--                FROM `albums` AS `LHS` -->
<!--                INNER JOIN `artists` AS `RHS` ON (`LHS`.`artistid` = `RHS`.`artistid`))) GROUP  BY `artistid`))) -->
<!-- WHERE (`num_albums` > 3.0) -->
<!--   ORDER  BY `num_albums` DESC -->
<!-- LIMIT 10; -->
<!-- ``` -->
<!-- ```{r} -->
<!-- # You can't see this, but output.var = "band_albums_sql" in the chunk above -->
<!-- # this. So we're outputting our data into band_albums_sql -->
<!-- band_albums_sql -->
<!-- ``` -->
<!-- SQL is more versatile than `dbplyr`, but it's also a lot more confusing because it's not as straightforward. This creates one golden rule: -->
<!-- > *Use dbplyr functions for data analysis and SQL code for accessing data* -->
<!-- This is because SQL code is weird and hard to understand, but it allows you to easily pull in data. SQL is a coding language *designed* for you to query or ask databases for data. Meanwhile, `dbplyr` is meant for analyzing that data and modifying it. -->
<!-- ```{r, include = FALSE} -->
<!-- dbDisconnect(con) -->
<!-- file.remove("chinook.db") -->
<!-- ``` -->
<!-- Now, we worked with an SQLite database within this tutorial due to its simplicity. However, there are other databases that can be accessed in different ways. For example, a PostgresSQL server is located on the internet and allows you to pull data from the internet instead of downloading it into your computer. You can learn about some of the types of SQL databases and their differences [here](https://www.digitalocean.com/community/tutorials/sqlite-vs-mysql-vs-postgresql-a-comparison-of-relational-database-management-systems). -->
<!-- *But why should we bother with databases at all?* -->
<!-- Opening a database stored in a `.db` file involved a lot more effort than just opening a `.csv`, `.tsv`, or any of the other plain text or Excel formats. It was a bit of a pain to use a database in that setting since we had to use `dbplyr` to translate `tidyverse`-like commands (`filter`, `select`, `head`, etc.) into SQL commands that the database understands. Not all `tidyverse` commands can currently be translated with SQLite databases, like we showed with `slice()`. So you might be wondering, why should we use databases at all?  -->
<!-- Databases are beneficial in a large-scale setting: -->
<!-- - they enable storing large data sets across multiple computers with automatic redundancy and backups -->
<!-- - they allow multiple users to access them simultaneously and remotely without conflicts and errors -->
<!-- - they provide mechanisms for ensuring data integrity and validating input -->
<!-- - they provide security to keep data safe -->
<!-- For example, [there are billions of Google searches conducted daily](https://www.internetlivestats.com/google-search-statistics/).  -->
<!-- This makes databases extremely useful for a business where they'll have employees doing thousands or millions of operations daily. This makes knowing what SQL databases are extremely vital from both a data science and a general usage perspective. -->
</section></section><section id="webscraping" class="level2" data-number="3.4"><h2 data-number="3.4" class="anchored" data-anchor-id="webscraping">
<span class="header-section-number">3.4</span> Webscraping</h2>
<p>Please read the <a href="https://r4ds.hadley.nz/webscraping.html">web scraping chapter</a> from <a href="https://r4ds.hadley.nz/"><em>R for Data Science (2e)</em></a>.</p>
</section><section id="working-with-apis" class="level2" data-number="3.5"><h2 data-number="3.5" class="anchored" data-anchor-id="working-with-apis">
<span class="header-section-number">3.5</span> Working with APIs</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="03-data/images/api.png" class="img-fluid figure-img" style="width:50.0%"></p>
</figure>
</div>
</div>
</div>
<!-- source used here: https://bookdown.org/rdpeng/RProgDA/reading-web-based-data.html -->
<p>“API” stands for Application Program Interface. They allow us to access open data from government agencies, companies, and other organizations. API provides the rules for software applications to interact with one another. Open data APIs provide the rules you need to know to write R code to request and pull data from the organization’s web server into R. Usually, some of the computational burden of querying and subsetting the data is taken on by the source’s server, to create the subset of requested data to pass to your computer. In practice, this means you can often pull the subset of data you want from a very large available dataset without having to download the full dataset and load it locally into your R session.</p>
<p>As an overview, the basic steps for accessing and using data from a web API when working in R are:</p>
<ul>
<li>Figure out the API rules for HTTP requests</li>
<li>Write R code to create a request in the proper format</li>
<li>Send the request using GET or POST HTTP methods</li>
<li>Once you get back data from the request, parse it into an easier-to-use format if necessary</li>
</ul>
<p>To get the data from an API, you should first read the organization’s API documentation. An organization will post details on what data is available through their API(s), as well as how to set up HTTP requests to get that data. To request the data through the API, you will typically need to send the organization’s web server an HTTP request using a GET or POST method. The API documentation details will typically show an example GET or POST request for the API, including the base URL to use and the possible query parameters that can be used to customize the dataset request.</p>
<p>Here is an example:</p>
<p>The National Aeronautics and Space Administration (NASA) has an API for pulling the <a href="https://apod.nasa.gov/apod/astropix.html">Astronomy Picture of the Day</a>. In their API documentation, they specify that the base URL for the API request should be <em>https://api.nasa.gov/planetary/apod</em> and that you can include parameters to specify the date of the daily picture you want, whether to pull a high-resolution version of the picture, and a NOAA API key you have requested from NOAA.</p>
<p>Many organizations will require you to get an <em>API key</em> and use this key in each of your API requests. This key allows the organization to control API access, including enforcing rate limits per user. API rate limits restrict how often you can request data (such as an hourly limit of 1,000 requests per user for NASA APIs).</p>
<p>API keys should be kept private, so if you are writing code that includes an API key, be very careful not to include the actual key in any code that is public (even any code in public GitHub repositories). To ensure privacy, save the value of your key in a file named .Renviron in your home directory. This file should be a plain text file and must end in a blank line. Once you’ve saved your API key to a global variable in that file (e.g., with a line added to the .Renviron file like NOAA_API_KEY = “abdafjsiopnab038”), you can assign the key value to an R object in an R session using the Sys.getenv function (e.g., noaa_api_key &lt;- Sys.getenv(“NOAA_API_KEY”)), and then use the object <code>noaa_api_key</code> anywhere you would otherwise have used the character string with your API key.</p>
<p>To find more R packages for accessing and exploring open data, check out the <a href="https://github.com/ropensci/opendata">Open Data CRAN task view</a>. You can also browse through the <a href="https://ropensci.org/packages/">ROpenSci packages</a>, all of which have GitHub repositories where you can further explore how each package works! ROpenSci is an organization with the mission to create open software tools for science. If you create your own package to access data relevant to scientific research through an API, consider submitting it for peer-review through ROpenSci.</p>
<p>The <strong>riem</strong> package, developed by Maelle Salmon and an ROpenSci package, is an excellent and straightforward example of how you can use R to pull open data through a web API. This package allows you to pull weather data from airports around the world directly from the Iowa Environmental Mesonet. To show you how to pull data into R through an API, in this section we will walk you through code in the <strong>riem</strong> package or code based closely on code in the package.</p>
<p>To get a certain set of weather data from the Iowa Environmental Mesonet, you can send an HTTP request specifying a base URL, <em>https://mesonet.agron.iastate.edu/cgi-bin/request/asos.py/</em>, as well as some parameters describing the subset of dataset you want (e.g., date ranges, weather variables, output format). Once you know the rules for the names and possible values of these parameters (more on that below), you can submit an HTTP GET request using the function<code><a href="https://httr.r-lib.org/reference/GET.html">GET()</a></code> from the <strong>httr</strong> package.</p>
<p>When you are making an HTTP request using the <code><a href="https://httr.r-lib.org/reference/GET.html">GET()</a></code> or <code><a href="https://httr.r-lib.org/reference/POST.html">POST()</a></code> functions from the <strong>httr</strong> package, you can include the key-value pairs for any query parameters as a list object in the <code>query</code> argument of the function. For example, suppose you want to get wind speed in miles per hour (data = “sped”) for Denver, CO, (station = “DEN”) for the month of June 2016 (year1 = “2016”, month1 = “6”, etc.) in Denver’s local time zone (tz = “America/Denver”) and in a comma-separated file (format = “comma”). To get this weather dataset, you can run:</p>
<div class="cell">
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://httr.r-lib.org/">httr</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://httr.r-lib.org/">httr</a></span><span class="op">)</span></span>
<span><span class="va">meso_url</span> <span class="op">&lt;-</span> <span class="st">"https://mesonet.agron.iastate.edu/cgi-bin/request/asos.py/"</span></span>
<span><span class="va">denver</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://httr.r-lib.org/reference/GET.html">GET</a></span><span class="op">(</span>url <span class="op">=</span> <span class="va">meso_url</span>,</span>
<span>                    query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>station <span class="op">=</span> <span class="st">"DEN"</span>,</span>
<span>                                 data <span class="op">=</span> <span class="st">"sped"</span>,</span>
<span>                                 year1 <span class="op">=</span> <span class="st">"2016"</span>,</span>
<span>                                 month1 <span class="op">=</span> <span class="st">"6"</span>,</span>
<span>                                 day1 <span class="op">=</span> <span class="st">"1"</span>,</span>
<span>                                 year2 <span class="op">=</span> <span class="st">"2016"</span>,</span>
<span>                                 month2 <span class="op">=</span> <span class="st">"6"</span>,</span>
<span>                                 day2 <span class="op">=</span> <span class="st">"30"</span>,</span>
<span>                                 tz <span class="op">=</span> <span class="st">"America/Denver"</span>,</span>
<span>                                 format <span class="op">=</span> <span class="st">"comma"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr.r-lib.org/reference/content.html">content</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span>skip <span class="op">=</span> <span class="fl">5</span>, na <span class="op">=</span> <span class="st">"M"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 9108 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (1): station
dbl  (1): sped
dttm (1): valid

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># There are 9,106 rows of data to look at! Let's just look at subset for our</span></span>
<span><span class="co"># purposes.</span></span>
<span></span>
<span><span class="va">denver</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 3
  station valid                sped
  &lt;chr&gt;   &lt;dttm&gt;              &lt;dbl&gt;
1 DEN     2016-06-01 00:00:00   9.2
2 DEN     2016-06-01 00:05:00   9.2
3 DEN     2016-06-01 00:10:00   6.9</code></pre>
</div>
</div>
<p>The <code><a href="https://httr.r-lib.org/reference/content.html">content()</a></code> call extracts the content from the response to the HTTP request sent by the <code><a href="https://httr.r-lib.org/reference/GET.html">GET()</a></code> function. The Iowa Environmental Mesonet API offers the option to return the requested data in a comma-separated file (format = “comma” in the GET request), so here content and <code><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv()</a></code> are used to extract and read in that csv file. Usually, data will be returned in a JSON format instead.</p>
<p>The only tricky part of this process is figuring out the available parameter names (e.g., station) and possible values for each (e.g., “DEN” for Denver). Currently, the details you can send in an HTTP request through Iowa Environmental Mesonet’s API include:</p>
<ul>
<li>A four-character weather station identifier (<code>station</code>)</li>
<li>The weather variables (e.g., temperature, wind speed) to include (<code>data</code>)</li>
<li>Starting and ending dates describing the range for which you’d like to pull data (<code>year1</code>, <code>month1</code>, <code>day1</code>, <code>year2</code>, <code>month2</code>, <code>day2</code>)</li>
<li>The time zone to use for date-times for the weather observations (<code>tz</code>)</li>
<li>Different formatting options (e.g., delimiter to use in the resulting data file [<code>format</code>], whether to include longitude and latitude)</li>
</ul>
<p>Typically, these parameter names and possible values are explained in the API documentation. In some cases, however, the documentation will be limited. In that case, you may be able to figure out possible values, especially if the API specifies a GET rather than POST method, by playing around with the website’s point-and-click interface and then looking at the url for the resulting data pages. For example, if you look at the <a href="https://mesonet.agron.iastate.edu/request/download.phtml?network=IN__ASOS">Iowa Environmental Mesonet’s page</a> for accessing this data, you’ll notice that the point-and-click web interface allows you the options in the list above, and if you click through to access a dataset using this interface, the web address of the data page includes these parameter names and values.</p>
<p>The <strong>riem</strong> package implements all these ideas in three very clean and straightforward functions. You can explore the code behind this package and see how these ideas can be incorporated into a small R package, in the /R directory of the <a href="https://github.com/ropensci/riem">package’s GitHub page</a>.</p>
<p>R packages already exist for many open data APIs. If an R package already exists for an API, you can use functions from that package directly, rather than writing your own code using the API protocols and httr functions. Other examples of existing R packages to interact with open data APIs include:</p>
<ol type="1">
<li>twitteR: Twitter</li>
<li>rnoaa: National Oceanic and Atmospheric Administration</li>
<li>Quandl: Quandl (financial data)</li>
<li>RGoogleAnalytics: Google Analytics</li>
<li>censusr, acs: United States Census</li>
<li>WDI, wbstats: World Bank</li>
<li>GuardianR, rdian: The Guardian Media Group</li>
<li>blsAPI: Bureau of Labor Statistics</li>
<li>rtimes: New York Times</li>
<li>dataRetrieval, waterData: United States Geological Survey If an R package doesn’t exist for an open API and you’d like to write your own package, find out more about writing API packages with <a href="https://cran.r-project.org/web/packages/httr/vignettes/api-packages.html">this vignette for the httr package</a>. This document includes advice on error handling within R code that accesses data through an open API.</li>
</ol>
<p>Information for this section on API’s was taken from <em>Mastering Software Development in R</em> textbook, authored by Roger D. Peng, Sean Kross, and Brooke Anderson.</p>
</section><section id="summary" class="level2" data-number="3.6"><h2 data-number="3.6" class="anchored" data-anchor-id="summary">
<span class="header-section-number">3.6</span> Summary</h2>
<!-- One paragraph for each subsection. Each paragraph tries to provide a brief summary, including package names and key commands. -->
<p><em>Pulling data into and out of R is a key factor in the data science process.</em></p>
<p><em>Use files like CSV, Excel, RDS, JSON, and SQL files/databases to organize and share your data.</em></p>
<p><em>Use SelectorGadget to pull data off of websites.</em></p>
<p><em>Use an API when to get data from government agencies or companies.</em></p>
<p>In this chapter, we looked at a few common file formats and used the <strong>readr</strong> package to read them and pull data from them so that we can use that data in our plots.</p>
<p>We also looked at databases and how we can pull in a few select pieces of data from them so that we aren’t overloading our computers with thousands or millions of data rows. We went over how we can write SQL queries so that we can acheive some special effects without causing errors or detonating our computers.</p>
<p>And finally, we looked at how we can pull data from websites both through webscraping and through APIs, letting us pull data from the internet quickly and easily. This lets us find data for our projects and load it into our R session without having to create it ourselves or download a file from the internet.</p>
<p>In the end, this chapter is about getting data from other people and using it inside your own projects.</p>
<div class="cell">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="03-data/images/suckville.png" class="img-fluid figure-img" width="308"></p>
<p></p><figcaption class="figure-caption">Make sure to use up to date data.</figcaption><p></p>
</figure>
</div>
</div>
</div>
<div class="cell">

</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./02-wrangling.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Wrangling</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./04-rubin-causal-model.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Rubin Causal Model</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>